rmw_zenoh_cpp/src/rmw_zenoh.cpp

```diff
@@ -928,65 +951,36 @@ rmw_create_subscription(
+  auto node_data = context_impl->get_node_data(node);
   RMW_CHECK_FOR_NULL_WITH_MSG(
-    sub_data,
-    "failed to allocate memory for subscription data",
+    node_data,
+    "NodeData not found.",
     return nullptr);
-  auto free_sub_data = rcpputils::make_scope_exit(
-    [sub_data, allocator]() {
-      allocator->deallocate(sub_data, allocator->state);
-    });
 
-  RMW_TRY_PLACEMENT_NEW(sub_data, sub_data, return nullptr, rmw_zenoh_cpp::rmw_subscription_data_t);
-  auto destruct_sub_data = rcpputils::make_scope_exit(
-    [sub_data]() {
-      RMW_TRY_DESTRUCTOR_FROM_WITHIN_FAILURE(
-        sub_data->~rmw_subscription_data_t(),
-        rmw_zenoh_cpp::rmw_subscription_data_t);
-    });
-
-  // Adapt any 'best available' QoS options
-  sub_data->adapted_qos_profile = *qos_profile;
-  rmw_ret_t ret = rmw_zenoh_cpp::QoS::get().best_available_qos(
-    node, topic_name, &sub_data->adapted_qos_profile, rmw_get_publishers_info_by_topic);
-  if (RMW_RET_OK != ret) {
-    RMW_SET_ERROR_MSG("Failed to obtain adapted_qos_profile.");
+  if (!node_data->create_sub_data(
+      rmw_subscription,
+      context_impl->session(),
+      context_impl->graph_cache(),
+      context_impl->get_next_entity_id(),
+      topic_name,
+      type_support,
+      qos_profile))
+  {
+    // Error already handled.
     return nullptr;
   }
 
-  sub_data->typesupport_identifier = type_support->typesupport_identifier;
-  sub_data->type_hash = type_support->get_type_hash_func(type_support);
-  sub_data->type_support_impl = type_support->data;
-  auto callbacks = static_cast<const message_type_support_callbacks_t *>(type_support->data);
-  sub_data->type_support = static_cast<rmw_zenoh_cpp::MessageTypeSupport *>(
-    allocator->allocate(sizeof(rmw_zenoh_cpp::MessageTypeSupport), allocator->state));
-  RMW_CHECK_FOR_NULL_WITH_MSG(
-    sub_data->type_support,
-    "Failed to allocate MessageTypeSupport",
-    return nullptr);
-  auto free_type_support = rcpputils::make_scope_exit(
-    [sub_data, allocator]() {
-      allocator->deallocate(sub_data->type_support, allocator->state);
-    });
-
-  RMW_TRY_PLACEMENT_NEW(
-    sub_data->type_support,
-    sub_data->type_support,
-    return nullptr,
-    rmw_zenoh_cpp::MessageTypeSupport, callbacks);
-  auto destruct_msg_type_support = rcpputils::make_scope_exit(
-    [sub_data]() {
-      RMW_TRY_DESTRUCTOR_FROM_WITHIN_FAILURE(
-        sub_data->type_support->~MessageTypeSupport(),
-        rmw_zenoh_cpp::MessageTypeSupport);
-    });
-
-  sub_data->context = node->context;
-
+  // Store type erased node in rmw_subscription->data so that the
+  // Subscription can be safely accessed.
+  // TODO(Yadunund): We cannot store the rmw_node_t * here since this type erased
+  // subscription handle will be returned in the rmw_subscriptions_t in rmw_wait
+  // from which we cannot obtain SubscriptionData.
+  // rmw_subscription->data = reinterpret_cast<void *>(const_cast<rmw_node_t *>(node));
```

Comment:
```suggestion
  // TODO(Yadunund): We cannot store the rmw_node_t * here since this type erased
  // subscription handle will be returned in the rmw_subscriptions_t in rmw_wait
  // from which we cannot obtain SubscriptionData.
```